src/shared/libmount-util.c

```diff
@@ -38,3 +42,58 @@ int libmount_parse(
         *ret_iter = TAKE_PTR(iter);
         return 0;
 }
+
+int path_is_temporary_fs_harder(const char *path) {
+        _cleanup_(mnt_free_tablep) struct libmnt_table *table = NULL;
+        _cleanup_(mnt_free_iterp) struct libmnt_iter *iter = NULL;
+        _cleanup_free_ char *upper = NULL;
+        int r;
+
+        /* Check if the specified path is on a temporary filesystem.
+         * This is like path_is_temporary_fs(), but supports overlayfs. If the mountpoint is found be an
+         * overlay, mount table is parsed to extract the upperdir= option for the mountpoint, and this
+         * mountpoint is checked recursively. This recursive check is done up to 5 times, in case somebody
+         * wants to go crazy and nest overlays.
+         */
+
+        for (unsigned tries = 0; tries < 5; tries++) {
```

Comment:
The upper bound `5` sounds too small for me. This is used by PID1, and if this fails, the PID1 will exit with an error. I think the bound should be higher for safety.

Reply:
Yeah, arbitrary constants are not nice. Should I make it something like 128 or 256?

Reply:
I added a `#define` and a comment and changed the limit to 128.